In [2]:
import numpy as np
import pandas as pd

## 1. Load raw data

In [3]:
df_raw = pd.read_csv('../data/raw/books_1.Best_Books_Ever.csv')

## 2. Basic Overview

In [4]:
df_raw.head(10)

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
0,2767052-the-hunger-games,The Hunger Games,The Hunger Games #1,Suzanne Collins,4.33,WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...,English,9780439023481,"['Young Adult', 'Fiction', 'Dystopia', 'Fantas...","['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...",...,NaN,['Locus Award Nominee for Best Young Adult Boo...,6376780,"['3444695', '1921313', '745221', '171994', '93...",96.0,"['District 12, Panem', 'Capitol, Panem', 'Pane...",https://i.gr-assets.com/images/S/compressed.ph...,2993816,30516,5.09
1,2.Harry_Potter_and_the_Order_of_the_Phoenix,Harry Potter and the Order of the Phoenix,Harry Potter #5,"J.K. Rowling, Mary GrandPré (Illustrator)",4.50,There is a door at the end of a silent corrido...,English,9780439358071,"['Fantasy', 'Young Adult', 'Fiction', 'Magic',...","['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...",...,06/21/03,['Bram Stoker Award for Works for Young Reader...,2507623,"['1593642', '637516', '222366', '39573', '14526']",98.0,['Hogwarts School of Witchcraft and Wizardry (...,https://i.gr-assets.com/images/S/compressed.ph...,2632233,26923,7.38
2,2657.To_Kill_a_Mockingbird,To Kill a Mockingbird,To Kill a Mockingbird,Harper Lee,4.28,The unforgettable novel of a childhood in a sl...,English,9999999999999,"['Classics', 'Fiction', 'Historical Fiction', ...","['Scout Finch', 'Atticus Finch', 'Jem Finch', ...",...,07/11/60,"['Pulitzer Prize for Fiction (1961)', 'Audie A...",4501075,"['2363896', '1333153', '573280', '149952', '80...",95.0,"['Maycomb, Alabama (United States)']",https://i.gr-assets.com/images/S/compressed.ph...,2269402,23328,NaN
3,1885.Pride_and_Prejudice,Pride and Prejudice,NaN,"Jane Austen, Anna Quindlen (Introduction)",4.26,Alternate cover edition of ISBN 9780679783268S...,English,9999999999999,"['Classics', 'Fiction', 'Romance', 'Historical...","['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...",...,01/28/13,[],2998241,"['1617567', '816659', '373311', '113934', '767...",94.0,"['United Kingdom', 'Derbyshire, England (Unite...",https://i.gr-assets.com/images/S/compressed.ph...,1983116,20452,NaN
4,41865.Twilight,Twilight,The Twilight Saga #1,Stephenie Meyer,3.60,About three things I was absolutely positive.\...,English,9780316015844,"['Young Adult', 'Fantasy', 'Romance', 'Vampire...","['Edward Cullen', 'Jacob Black', 'Laurent', 'R...",...,10/05/05,"['Georgia Peach Book Award (2007)', 'Buxtehude...",4964519,"['1751460', '1113682', '1008686', '542017', '5...",78.0,"['Forks, Washington (United States)', 'Phoenix...",https://i.gr-assets.com/images/S/compressed.ph...,1459448,14874,2.1
5,19063.The_Book_Thief,The Book Thief,NaN,Markus Zusak (Goodreads Author),4.37,Librarian's note: An alternate cover edition c...,English,9780375831003,"['Historical Fiction', 'Fiction', 'Young Adult...","['Liesel Meminger', 'Hans Hubermann', 'Rudy St...",...,09/01/05,['National Jewish Book Award for Children’s an...,1834276,"['1048230', '524674', '186297', '48864', '26211']",96.0,"['Molching (Germany)', 'Germany']",https://i.gr-assets.com/images/S/compressed.ph...,1372809,14168,3.8
6,170448.Animal_Farm,Animal Farm,NaN,"George Orwell, Russell Baker (Preface), C.M. W...",3.95,Librarian's note: There is an Alternate Cover ...,English,9780451526342,"['Classics', 'Fiction', 'Dystopia', 'Fantasy',...","['Snowball', 'Napoleon', 'Clover', 'Boxer', 'O...",...,08/17/45,"['Prometheus Hall of Fame Award (2011)', 'Retr...",2740713,"['986764', '958699', '545475', '165093', '84682']",91.0,"['England', 'United Kingdom']",https://i.gr-assets.com/images/S/compressed.ph...,1276599,13264,4.42
7,11127.The_Chronicles_of_Narnia,The Chronicles of Narnia,The Chronicles of Narnia (Publication Order) #1–7,"C.S. Lewis, Pauline Baynes (Illustrator)",4.26,"Journeys to the end of the world, fantastic cr...",English,9999999999999,"['Fantasy', 'Classics', 'Fiction', 'Youn

## 3. Summary Statistics

In [5]:
df_raw.shape

(52478, 25)

### 3.1 Numeric column summary

In [6]:
df_raw.describe()

,rating,numRatings,likedPercent,bbeScore,bbeVotes
count,52478.000000,5.247800e+04,51856.000000,5.247800e+04,52478.000000
mean,4.021878,1.787865e+04,92.231545,1.984023e+03,22.529003
std,0.367146,1.039448e+05,5.990689,3.515314e+04,369.158541
min,0.000000,0.000000e+00,0.000000,0.000000e+00,-4.000000
25%,3.820000,3.410000e+02,90.000000,8.400000e+01,1.000000
50%,4.030000,2.307000e+03,94.000000,9.700000e+01,1.000000
75%,4.230000,9.380500e+03,96.000000,1.870000e+02,2.000000
max,5.000000,7.048471e+06,100.000000,2.993816e+06,30516.000000


Initial observations:
- Numeric columns include rating, number of ratings, liked percentage, BBE score, and BBE votes.
- Rating appears to be bounded between 0 and 5.
- Rating has zero scew, numRatings, bbeScore and bbeVotes has a positive skew, likedPercent has a very slight negative scew.
- The high scewness might be due to higher positive reviews or people prefering to suggest the books they like.
- bbeVotes (Number of votes in Best Books Ever list) has imposible value of -4, which need further investigation before analysis.
- numRatings, bbeScore(Score in Best Books Ever list), bbeVotes (Number of votes in Best Books Ever list) has outiliers, due to a large difference between 75% procentile and max value of the columns
- Minimum values of 0 could be due to missing or placeholder data as the list is voted on my general Goodreads community.

### 3.2 Text/object column summary

In [7]:
df_raw.describe(include="str")

,bookId,title,series,author,description,language,isbn,genres,characters,bookFormat,edition,pages,publisher,publishDate,firstPublishDate,awards,ratingsByStars,setting,coverImg,price
count,52478,52478,23470,52478,51140,48672,52478,52478,52478,51005,4955,50131,48782,51598,31152,52478,52478,52478,51873,38113
unique,52424,49927,22802,28227,50888,81,48072,44154,12448,136,1788,1365,11110,10635,8024,9215,49908,4651,51819,3776
top,8794263-promises-to-keep,Legacy,Star Wars Legends,Nora Roberts (Goodreads Author),"هذه هي طبعة ""دار الفكر - بيروت"" وهي آخر طبعة ع...",English,9999999999999,[],[],Paperback,First Edition,320,Vintage,Published,01/01/12,[],[],[],https://i.gr-assets.com/images/S/compressed.ph...,5.20
freq,2,14,15,86,37,42661,4354,4623,38712,26773,458,1049,572,586,225,41864,1423,40900,2,626


| Column           | Current type  | Real meaning                   |
| ---------------- | ------------- | ------------------------------ |
|`author`          | object/string | mixed author/contributor field |
|`language`        | object/string | categorical                    |
|`genres`          | object/string | list-like                      |
|`characters`      | object/string | list-like                      |
|`bookFormat`      | object/string | categorical                    |
|`pages`           | object/string | numeric-like                   |
|`publishDate`     | object/string | date-like                      |
|`firstPublishDate`| object/string | date-like                      |
|`awards`          | object/string | list-like                      |
|`ratingsByStars`  | object/string | list-like numeric counts       |
|`setting`         | object/string | list-like                      |
|`price`           | object/string | numeric-like                   |





Initial observations:
- Many object columns are not clean categorical variables.
- Some columns contain list-like strings, such as genres, characters, awards, ratingsByStars, and setting.
- Some columns are numeric-like or date-like but currently stored as text, such as pages, price, publishDate, and firstPublishDate.
- These columns need dedicated cleaning decisions before EDA.

### 3.3 Column role clasification
| Role                    |Columns                             | Notes                      |
| ----------------------- | ---------------------------------- | ---------------------------|
| Identifiers             | `bookId`, `isbn`                                                                 | `bookId` and `isbn` are mainly useful for tracking records, validating uniqueness, and investigating duplicates. They should not be treated as analytical features without additional validation. |
| Core metadata           | `title`, `author`, `language`, `bookFormat`, `publisher`, `series`, `edition`    | Core metadata columns such describe the book record. These columns are useful for filtering, grouping, and understanding duplicate records. Some of them require cleaning.  |
| Reader-response metrics | `rating`, `numRatings`, `likedPercent`, `ratingsByStars`, `bbeScore`, `bbeVotes` | These are reader-response or popularity-related columns. These are central to the project because they help define book success. However, they measure different things, and some need to be considered together to avoid overvaluing books.   |
| Temporal fields         | `publishDate`,`firstPublishDate`                                                 | They are currently stored as text and may contain inconsistent or ambiguous formats. For EDA, extracting publication year may be more reliable than trying to use full dates.                                  |
| List-like fields        | `genres`, `characters`, `awards`,`setting`                                       |  These are list-like columns stored as text. They need parsing before analysis. `genres` is high priority because it supports genre analysis and dashboard filters.  |
| Text/NLP field          |`description`                                                                     | description is a free-text column and is the main input for the NLP layer. It can be used for embeddings, clustering, semantic similarity, and recommendation features. However, it describes book content or marketing summaries, not actual reader sentiment, so it should not be interpreted as reader opinion.       |
| Numeric-like attributes | `pages`,`price`                                                                  | These are numeric-like attributes but currently are stored as text. `pages` can support analysis of book length versus rating/popularity.   |
| Media/link metadata     |`coverImg`                                                                        | Useful for dashboard display, not core analysis.                                   |


## 4. Schema Inspection

In [8]:
df_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 52478 entries, 0 to 52477
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   bookId            52478 non-null  str    
 1   title             52478 non-null  str    
 2   series            23470 non-null  str    
 3   author            52478 non-null  str    
 4   rating            52478 non-null  float64
 5   description       51140 non-null  str    
 6   language          48672 non-null  str    
 7   isbn              52478 non-null  str    
 8   genres            52478 non-null  str    
 9   characters        52478 non-null  str    
 10  bookFormat        51005 non-null  str    
 11  edition           4955 non-null   str    
 12  pages             50131 non-null  str    
 13  publisher         48782 non-null  str    
 14  publishDate       51598 non-null  str    
 15  firstPublishDate  31152 non-null  str    
 16  awards            52478 non-null  str    
 17  numR

The dataset contains 52478 rows and 25 columns. 20 columns are currenty stored as string but may contain categorical, list or numeric values and 5 are numeric. Book metadata columns like title, series are high-cardinality columns. `bookId` is the ID column of the dataset.

## 5. Missingness audit

In [9]:
missing = df_raw.isnull().mean().sort_values(ascending=False).mul(100).round(2)
missing

edition             90.56
series              55.28
firstPublishDate    40.64
price               27.37
language             7.25
publisher            7.04
pages                4.47
bookFormat           2.81
description          2.55
publishDate          1.68
likedPercent         1.19
coverImg             1.15
isbn                 0.00
ratingsByStars       0.00
bbeVotes             0.00
bbeScore             0.00
setting              0.00
awards               0.00
numRatings           0.00
genres               0.00
author               0.00
rating               0.00
title                0.00
characters           0.00
bookId               0.00
dtype: float64

The highest missingness appears in `edition`, `series`, `firstPublishDate`, and `price`.

These missing values should not be handled the same way:
- `edition` is likely optional metadata and may not be useful for core analysis.
- `series` missingness may be meaningful because many books are standalone.
- `firstPublishDate` is important for temporal analysis but has substantial missingness.
- `price` needs further investigation before deciding whether to use it.

## 6. Duplicate Audit

In [10]:
exact_duplicate_count = df_raw.duplicated().sum()
exact_duplicate_count

np.int64(50)

In [11]:
exact_duplicates = df_raw[df_raw.duplicated(keep=False)].sort_values("bookId")
exact_duplicates.head(10)

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
37460,11737313-three-times-lucky,Three Times Lucky,Mo & Dale Mysteries #1,Sheila Turnage,4.02,"Newbery honor winner, New York Times bestselle...",English,9780803736702,"['Mystery', 'Middle Grade', 'Fiction', 'Realis...",[],...,NaN,"['Newbery Medal Nominee (2013)', ""Pennsylvania...",22422,"['8633', '7941', '4181', '1076', '591']",93.0,['North Carolina (United States)'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,3.06
37406,11737313-three-times-lucky,Three Times Lucky,Mo & Dale Mysteries #1,Sheila Turnage,4.02,"Newbery honor winner, New York Times bestselle...",English,9780803736702,"['Mystery', 'Middle Grade', 'Fiction', 'Realis...",[],...,NaN,"['Newbery Medal Nominee (2013)', ""Pennsylvania...",22422,"['8633', '7941', '4181', '1076', '591']",93.0,['North Carolina (United States)'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,3.06
37442,1258333.Cantarella_Volume_1,"Cantarella, Volume 1",Cantarella #1,You Higuri,3.81,"From birth, Cesare Borgia is surrounded by sha...",English,9780976895701,"['Manga', 'Historical Fiction', 'Fantasy', 'Gr...","['Cesare Borgia', 'Lucrezia Borgia', 'Pope Ale...",...,03/08/01,[],933,"['324', '250', '253', '71', '35']",89.0,['Italy'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.67
37388,1258333.Cantarella_Volume_1,"Cantarella, Volume 1",Cantarella #1,You Higuri,3.81,"From birth, Cesare Borgia is surrounded by sha...",English,9780976895701,"['Manga', 'Historical Fiction', 'Fantasy', 'Gr...","['Cesare Borgia', 'Lucrezia Borgia', 'Pope Ale...",...,03/08/01,[],933,"['324', '250', '253', '71', '35']",89.0,['Italy'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.67
37428,13571757-1,アオハライド 1,アオハライド [Ao Haru Ride] #1,"Io Sakisaka, 咲坂伊緒",4.33,高1の双葉には中学時代から忘れられない初恋の人、「田中くん」がいる。中学のとき女子に、はぶら...,Japanese,9784088466477,"['Manga', 'Romance', 'Young Adult', 'Shojo', '...",[],...,NaN,[],16361,"['9121', '4447', '2102', '451', '240']",96.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,13.30
37482,13571757-1,アオハライド 1,アオハライド [Ao Haru Ride] #1,"Io Sakisaka, 咲坂伊緒",4.33,高1の双葉には中学時代から忘れられない初恋の人、「田中くん」がいる。中学のとき女子に、はぶら...,Japanese,9784088466477,"['Manga', 'Romance', 'Young Adult', 'Shojo', '...",[],...,NaN,[],16361,"['9121', '4447', '2102', '451', '240']",96.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,13.30
37461,13614864-know-not-why,Know Not Why,Know Not Why #1,Hannah Johnson (Goodreads Author),4.02,Howie gets a job at Artie Kraft's Arts 'N Craf...,English,9781476149035,"['M M Romance', 'LGBT', 'Romance', 'Contempora...","[""Howard 'Howie' Jenkins"", 'Arthur Kraft Jr.',...",...,NaN,[],3411,"['1290', '1231', '632', '189', '69']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37407,13614864-know-not-why,Know Not Why,Know Not Why #1,Hannah Johnson (Goodreads Author),4.02,Howie gets a job at Artie Kraft's Arts 'N Craf...,English,9781476149035,"['M M Romance', 'LGBT', 'Romance', 'Contempora...","[""Howard 'Howie' Jenkins"", 'Arthur Kraft Jr.',...",...,NaN,[],3411,"['1290', '1231', '632', '189', '69']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37483,1481056.Gifts_and_Ministries_of_the_Holy_Spirit,Gifts and Ministries of the Holy Spirit,NaN,Lester Sumrall,4.44,You Can Have the Power of the Holy Spirit The...,English,9780883686522,"['Christian', 'Theology', 'Christian Non Ficti...",[],...,06/06/77,[],333,"['206', '83', '31', '11', '2']",96.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,5.00
37429,1481056.Gifts_and_Ministries_of_the_Holy_Spirit,Gifts and Ministries of the Holy Spirit,NaN,Lester Sumrall,4.44,You Can Have the Power of the Holy Spirit The...,English,9780883686522,"['Christian', 'Theology', 'Christian Non Ficti...",[],...,06/06/77,[],333,"['206', '83', '31', '11', '2']",96.0,[],https://i.gr-assets.com/images/S/

In [12]:
bookid_duplicate_count = df_raw.duplicated(subset=["bookId"]).sum()
bookid_duplicate_count

np.int64(54)

In [ ]:
title_author_duplicates = df_raw[ df_raw.duplicated(subset=["title", "author"], keep=False) ].sort_values(["title", "author"])

title_author_duplicates.head(20)

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
2876,6443834-1q84,1Q84,1Q84 #1,"Haruki Murakami, 村上 春樹",4.00,1949年にジョージ・オーウェルは、近未来小説としての『1984』を刊行した。\nそして20...,Japanese,9784103534228,"['Fiction', 'Fantasy', 'Japan', 'Japanese Lite...",[],...,NaN,[],49121,"['17854', '18852', '8553', '2421', '1441']",92.0,['Tokyo (Japan)'],https://i.gr-assets.com/images/S/compressed.ph...,1218,15,9.09
4463,7717482-1q84,1Q84,1Q84 #3,"Haruki Murakami, 村上 春樹",3.90,1949年にジョージ・オーウェルは、近未来小説としての『1984』を刊行した。\n\nそして...,Japanese,9784103534259,"['Fiction', 'Fantasy', 'Japan', 'Japanese Lite...",[],...,10/28/10,[],28602,"['9232', '10355', '6565', '1952', '498']",91.0,['Tokyo (Japan)'],https://i.gr-assets.com/images/S/compressed.ph...,682,9,13.93
1679,12177850-a-song-of-ice-and-fire,A Song of Ice and Fire,A Song of Ice and Fire #1-5,George R.R. Martin,4.56,"For the first time, all five novels in the epi...",English,9781780484259,"['Fantasy', 'Fiction', 'Science Fiction Fantas...","['Tyrion Lannister', 'Arya Stark', 'Khal Drogo...",...,10/28/00,[],53233,"['36854', '11677', '3118', '890', '694']",97.0,[],https://i.gr-assets.com/images/S/compressed.ph...,2616,35,NaN
2899,9814682-a-song-of-ice-and-fire,A Song of Ice and Fire,A Song of Ice and Fire #1-4,George R.R. Martin,4.63,George R. R. Martin's A Song of Ice and Fire s...,English,9780345529053,"['Fantasy', 'Fiction', 'Epic Fantasy', 'Advent...",[],...,01/01/11,[],44894,"['32256', '9608', '2253', '506', '271']",98.0,[],https://i.gr-assets.com/images/S/compressed.ph...,1201,16,12.24
37383,24633605-always-and-forever,Always and Forever,Serenity Point #2,Harper Bentley (Goodreads Author),3.93,Does wanting to slap the hell out of Brody Kel...,English,9999999999999,"['Contemporary Romance', 'Romance', 'Firefight...",[],...,NaN,[],482,"['153', '188', '104', '26', '11']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37437,24633605-always-and-forever,Always and Forever,Serenity Point #2,Harper Bentley (Goodreads Author),3.93,Does wanting to slap the hell out of Brody Kel...,English,9999999999999,"['Contemporary Romance', 'Romance', 'Firefight...",[],...,NaN,[],482,"['153', '188', '104', '26', '11']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37412,3973924-bal-masqu,Bal masqué,NaN,"Elia Barceló, Dorotea ter Horst (Translator)",3.79,"Parijs, 1991. Jaren nadat de Argentijnse schri...",Dutch,9789021453293,"['Roman', 'Fiction']",[],...,01/01/04,[],341,"['77', '146', '91', '24', '3']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37466,3973924-bal-masqu,Bal masqué,NaN,"Elia Barceló, Dorotea ter Horst (Translator)",3.79,"Parijs, 1991. Jaren nadat de Argentijnse schri...",Dutch,9789021453293,"['Roman', 'Fiction']",[],...,01/01/04,[],341,"['77', '146', '91', '24', '3']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37426,26845691-beatlebone,Beatlebone,NaN,Kevin Barry,3.47,He will spend three days alone on his island. ...,English,9781782116134,"['Fiction', 'Ireland', 'Irish Literature', 'Mu...",[],...,NaN,['Goldsmiths Prize (2015)'],2528,"['471', '860', '729', '334', '134']",81.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.16
37480,26845691-beatlebone,Beatlebone,NaN,Kevin Barry,3.47,He will spend three days alone on his island. ...,English,9781782116134,"['Fiction', 'Ireland', 'Irish Literature', 'Mu...",[],...,NaN,['Goldsmiths Prize (2015)'],2528,"['471', '860', '729', '334', '134']",81.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.16


Exact duplicate rows can be removed during cleaning because they repeat the same record.

However, `title + author` duplicates require caution. These may represent different editions, formats, translations, or duplicate scraped records. They should not be automatically dropped without further investigation.

## 7. Dataset Grain and Duplicate Strategy

Before cleaning or removing duplicate records, I need to define what one row represents. In this dataset, one row appears to represent a Goodreads book record or edition-level entry, not necessarily one unique literary work. This distinction matters because the same book title and author may appear multiple times due to different editions, formats, translations, publication dates, or ISBNs.

| Possible grain     | Meaning                                                | Consequence                                 |
| ------------------ | ------------------------------------------------------ | ------------------------------------------- |
| Record-level       | Each row is one scraped Goodreads record               | Safest assumption for now                   |
| Edition-level      | Each row is a specific edition or format of a book     | Same title/author can appear multiple times |
| Work-level         | Each row is one unique book/work regardless of edition | Would require merging editions              |
| Author-title-level | Each row is one title-author combination               | Too simplistic and risky                    |


I’ll be concentrating on handling the dataset as individual records or editions.


### 7.1 Duplicate type 1: exact full-row duplicates

In [14]:
exact_duplicate_count = df_raw.duplicated().sum()
exact_duplicate_count

np.int64(50)

Exact duplicate rows likely represent repeated scraped records from Goodreads list. There are 50 excact duplicates. Since all values are identical, removing them should not lose unique information.

### 7.2 Duplicate type 2: duplicate `bookId`

In [15]:
bookid_duplicate_count = df_raw.duplicated(subset=["bookId"]).sum()
bookid_duplicate_count

np.int64(54)

Duplicate `bookId` values are suspicious because `bookId` should identify a unique source record. These records require inspection to determine whether they are exact duplicates, near-duplicates, or inconsistent entries. An overview of the data seems to show that they are exact duplicates most of the time.

### 7.3 Duplicate type 3: duplicate `isbn`

In [18]:
isbn_duplicate = df_raw[ df_raw.duplicated(subset=["isbn"], keep=False)]
isbn_duplicate.head(10)

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
2,2657.To_Kill_a_Mockingbird,To Kill a Mockingbird,To Kill a Mockingbird,Harper Lee,4.28,The unforgettable novel of a childhood in a sl...,English,9999999999999,"['Classics', 'Fiction', 'Historical Fiction', ...","['Scout Finch', 'Atticus Finch', 'Jem Finch', ...",...,07/11/60,"['Pulitzer Prize for Fiction (1961)', 'Audie A...",4501075,"['2363896', '1333153', '573280', '149952', '80...",95.0,"['Maycomb, Alabama (United States)']",https://i.gr-assets.com/images/S/compressed.ph...,2269402,23328,NaN
3,1885.Pride_and_Prejudice,Pride and Prejudice,NaN,"Jane Austen, Anna Quindlen (Introduction)",4.26,Alternate cover edition of ISBN 9780679783268S...,English,9999999999999,"['Classics', 'Fiction', 'Romance', 'Historical...","['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...",...,01/28/13,[],2998241,"['1617567', '816659', '373311', '113934', '767...",94.0,"['United Kingdom', 'Derbyshire, England (Unite...",https://i.gr-assets.com/images/S/compressed.ph...,1983116,20452,NaN
7,11127.The_Chronicles_of_Narnia,The Chronicles of Narnia,The Chronicles of Narnia (Publication Order) #1–7,"C.S. Lewis, Pauline Baynes (Illustrator)",4.26,"Journeys to the end of the world, fantastic cr...",English,9999999999999,"['Fantasy', 'Classics', 'Fiction', 'Young Adul...","['Polly', 'Aslan', 'Lucy Pevensie', 'Edmund Pe...",...,10/28/56,[],517740,"['254964', '167572', '74362', '15423', '5419']",96.0,"['London, England']",https://i.gr-assets.com/images/S/compressed.ph...,1238556,12949,NaN
10,11870085-the-fault-in-our-stars,The Fault in Our Stars,NaN,John Green (Goodreads Author),4.21,Despite the tumor-shrinking medical miracle th...,English,9999999999999,"['Young Adult', 'Romance', 'Fiction', 'Contemp...","['Hazel Grace Lancaster', 'Augustus Waters', '...",...,NaN,"['Georgia Peach Book Award (2013)', 'Buxtehude...",3550714,"['1784471', '1022406', '512574', '150365', '80...",93.0,"['Indianapolis, Indiana (United States)', 'Ams...",https://i.gr-assets.com/images/S/compressed.ph...,1087056,11287,NaN
11,386162.The_Hitchhiker_s_Guide_to_the_Galaxy,The Hitchhiker's Guide to the Galaxy,The Hitchhiker's Guide to the Galaxy #1,Douglas Adams,4.22,Seconds before the Earth is demolished to make...,English,9999999999999,"['Science Fiction', 'Fiction', 'Humor', 'Fanta...","['Zaphod Beeblebrox', 'Arthur Dent', 'Ford Pre...",...,10/12/79,[],1436325,"['725771', '420864', '199846', '58326', '31518']",94.0,[],https://i.gr-assets.com/images/S/compressed.ph...,1063601,10996,NaN
14,968.The_Da_Vinci_Code,The Da Vinci Code,Robert Langdon #2,Dan Brown (Goodreads Author),3.86,ISBN 9780307277671 moved to this edition.While...,English,9999999999999,"['Fiction', 'Mystery', 'Thriller', 'Suspense',...","['Sophie Neveu', 'Robert Langdon', 'Sir Leigh ...",...,03/18/03,['British Book Award for Book of the Year (200...,1933446,"['645308', '667657', '399278', '142103', '79100']",89.0,"['Paris (France)', 'London, England', 'France'...",https://i.gr-assets.com/images/S/compressed.ph...,876633,9231,NaN
16,5297.The_Picture_of_Dorian_Gray,The Picture of Dorian Gray,NaN,"Oscar Wilde, Jeffrey Eugenides (Introduction)",4.08,"Written in his distinctively dazzling manner, ...",English,9999999999999,"['Classics', 'Fiction', 'Horror', 'Fantasy', '...","['Basil Hallward', 'Lord Henry Wotton', 'Sibyl...",...,07/20/90,[],966196,"['382985', '350778', '176188', '41869', '14376']",94.0,"['England', 'England']",https://i.gr-assets.com/images/S/compressed.ph...,861799,9096,NaN
19,24280.Les_Mis_rables,Les Misérables,NaN,"Victor Hugo, Lee Fahnestock (Translator), Norm...",4.18,Introducing one of the most famous characters ...,English,9999999999999,"['Classics', 'Fiction', 'Historical Fiction', ...","['Jean Valjean', 'Javert', 'Cosette', 'Fantine...",...,10/28/62,[],682793,"['332692', '202012', '100932', '29673', '17484']",93.0,['Pari

Duplicate ISBN values may indicate duplicate edition records, but ISBN analysis should be inspected carefully because some ISBN values are missing, placeholder-like, or malformed.

ISBN duplicate checks are preliminary because ISBN cleaning has not been done yet.

### 7.4 Duplicate type 4: duplicate `title + author`

In [20]:
len(title_author_duplicates)

171

Duplicate `title + author` combinations should not be automatically removed. They may represent different editions, formats, translations, or publication records and the number might change after removing the exact duplicates from the dataset. These rows are important for understanding the dataset grain.

### 7.5 Duplicate type 5: duplicate normalized title only

Duplicate titles alone are not reliable indicators of duplicate books because different books can share the same or similar titles.

## 8. Cleaning Decision Table

| Column / area      | Raw problem                                                      | Cleaning decision                                  | Output / result                                  | Priority     |
| ------------------ | ---------------------------------------------------------------- | -------------------------------------------------- | ------------------------------------------------ | ------------ |
| Whole dataset      | Exact duplicate rows exist                                       | Remove exact duplicates only                       | Cleaner base dataset                             | High         |
| `bookId`           | Some duplicate IDs                                               | Investigate before dropping                        | Decide whether duplicate IDs are true duplicates | High         |
| `title + author`   | Many duplicates                                                  | Keep for now; likely editions/formats/translations | No automatic deletion                            | High         |
| `author`           | Main authors mixed with illustrators, translators, introductions | Separate primary authors from contributors         | New column `primary_authors`, `contributors`     | High         |
| `genres`           | Stored as stringified list                                       | Parse into real list                               | `genres_list`                                    | High         |
| `ratingsByStars`   | Stored as list-like string                                       | Convert into five numeric star-count columns       | `star_5`, `star_4`, `star_3`, `star_2`, `star_1` | High         |
| `publishDate`      | Messy formats and two-digit years                                | Extract year, flag parse quality                   | `publish_year`, `publish_date_quality`           | High         |
| `firstPublishDate` | Important but many missing values                                | Extract year, preserve missingness                 | `first_publish_year`                             | High         |
| `rating`           | Numeric but needs validity check                                 | Validate range 0–5                                 | Keep as numeric                                  | High         |
| `numRatings`       | Popularity measure, may have extreme skew                        | Validate non-negative, maybe log-transform later   | `numRatings`, later `log_numRatings`             | High         |
| `bbeScore`         | Ranking/popularity feature                                       | Validate and preserve                              | Keep raw, maybe normalize later                  | High         |
| `bbeVotes`         | May contain invalid negative values                              | Investigate negative values before using           | Cleaned or flagged value                         | High         |
| `description`      | Needed for NLP; may contain missing/short text                   | Audit length, language, duplicates, noise          | `description_clean`, `description_length`        | High for NLP |
| `isbn`             | Contain placeholders and maybe malformed values                  | Keep raw ISBN, create cleaned ISBN validity flag   | New column `isbn_valid`                          | Medium       |
| `awards`           | Stored as stringified list                                       | Parse; create award count and award flag           | `awards_list`, `award_count`, `has_award`        | Medium       |
| `pages`            | Numeric-like but may contain invalid values                      | Convert to numeric, flag bad values                | `pages_num`, `pages_valid`                       | Medium       |
| `likedPercent`     | Percentage field                                                 | Validate 0–100                                     | Keep or convert to proportion                    | Medium       |
| `language`         | Many unique values, possible inconsistency                       | Normalize categories; group rare languages         | `language_clean`                                 | Medium       |
| `bookFormat`       | Explains edition duplicates                                      | Standardize values                                 | `book_format_clean`                              | Medium       |
| `series`           | Missingness may mean standalone book                             | Create `is_series` flag                            | `is_series`                                      | Medium       |
| `price`            | High missingness and unclear meaning                             | Parse only if needed later                         | `price_parsed`                                   | Low/Medium   |
| `characters`       | Stored as stringified list;                                      | Parse only if needed later                         | `characters_list`                                | Low          |
| `setting`          | Stored as stringified list;                                      | Parse but do not prioritize for first EDA          | `setting_list`                                   | Low          |
| `edition`          | Very high missingness                                            | Parse only if needed later                         | `edition_parsed`                                 | Low          |

## 8. Fake Missing / Placeholder Value Audit

The previous missingness audit only checks formal null values. However, some fields may contain placeholder values, empty list strings, invalid numbers, or text labels that represent missing or unusable information. This section checks for values that are technically present but may need to be treated as missing, invalid, or low-quality during cleaning.

## 9. Validity Checks Before Cleaning